source/mothership-design.md

```diff
@@ -0,0 +1,771 @@
++--------------------+-----------------+------------------+------------------+
+| `CMND`, `RUN`      | `READY`         |                  | `RUNNING`        |
++--------------------+-----------------+------------------+------------------+
+| `CMND`, `STOP`     | `RUNNING`       | `STOPPING`       | `STOPPED`        |
++--------------------+-----------------+------------------+------------------+
+| `CMND`, `RECL`     | `UNDERDEFINED`  |                  | None             |
+|                    | or `DEFINED` or |                  |                  |
+|                    | `READY` or      |                  |                  |
+|                    | `STOPPED`       |                  |                  |
++--------------------+-----------------+------------------+------------------+
+
+Table: Input key permutations, and how they change the state of an application
+on the Mothership (assuming the operations succeed). Note that C&C messages
+processed before the application reaches the input state are "stored", and
+enacted when the application reaches the input state via some other C&C
+message. Transition states exist to aid loading and stopping "completion
+detection" by `DebugInputBroker` (see the Debugging section).
+
+[^last]: This state is only set when the final message is received (see the
+    Command and Control section for more information on `APP` messages).
+
+[^none]: "None" here means that `AppDB` and `SBase` do not know about the task,
+    and do not hold any information on it.
+
+# Supervisors
+Supervisors are devices in an application that exist on the Mothership, and can
+communicate with normal devices serviced by the box associated with its
+Mothership, as well as external devices elsewhere. They are:
+
+ - Defined in the application description (XML)
+
+ - Compiled into shared object files by `P_Builder`
+
+ - Deployed to the Mothership process via an (`APP`, `SUPD`) message
+
+ - Loaded as part of an application by the Mothership process when the (`CMND`,
+   `INIT`) message is received.
+
+ - Represented by a `SuperHolder` object with the following fields (defined on
+   construction - a supervisor that cannot be loaded and is not fully defined
+   will cause the Mothership to report and set the application state to
+   `BROKEN`):
+
+     - `std::string path`: Where the supervisor was loaded from.
+
+     - `void* so`: The dynamically-loaded supervisor (using `dlopen`), which
+       populates the function pointer members.
+
+     - `int (*initialise)()`: Called when the application is started.
+
+     - `int (*entryPoint)(PMsg_p, PMsg_p)`: The entry point for all supervisor
+       calls while the application is running.
+
+ - Stored in the `SuperDB` object (`Mothership.superdb`) within
+   `std::map<std::string, SuperHolder> SuperDB.supervisors`, keyed by
+   application name. For an incoming packet, the appropriate supervisor is
+   identified from the task component of the software address.
```

Comment:
What happens if a message is received for a task that does not exist? :sweet_potato:

Reply:
An error (515) is posted (added a sentence).